### ISSUE TITLE
Missing opus support

### ISSUE THREAD

**Issue by badaix** [badaix/snapweb#8]
Snapweb should support opus. There is a [opus branch](https://github.com/badaix/snapweb/tree/opus) with a emscripten cross compiled opus library, but it's getting out of sync during playback, maybe the decoding must be done in a thread.

**Comment by tilllt** [comment]
Naive question, but can't snapweb use the built-in browsers codecs (like OPUS) and achieve sync by using browser internal mechanisms like this?

https://developer.mozilla.org/en-US/docs/Web/API/AudioBufferSourceNode/playbackRate

**Comment by kingosticks** [comment]
No. We need to be able to decode raw Opus, browsers handle Ogg Opus. I don't think playbackrate helps us with this but maybe I'm misunderstanding you. Something like https://github.com/samirkumardas/opus-to-pcm goes some way to explain the issue (I'm not an expert on Ogg but I'm not convinced about the framing implementation and the flushing config setting). More interesting is the upcoming https://w3c.github.io/webcodecs/ API which would be perfect. Looks like it's in Chrome 86 (behind a feature flag) and Firefox are working on supporting it. I guess it won't be available for a long while yet.

**Comment by tilllt** [comment]
I understand that snapcast needs the raw opus stream, not the ogg encapsuled opus, so that's a problem if the built-in codec cannot decode raw opus.

  In any case, it seems that other programs successfully integrated emscripten versions of opus, specifically the aurora.js project...

https://github.com/audiocogs/opus.js

**Comment by kingosticks** [comment]
Yes, someone just needs to do it.

**Comment by m040601** [comment]
> Snapweb should support opus. 

After some years of using snapcast already, I ended up here today because I was testing the builtin Snapweb client on Firefox (http://myserver:1780/).

I rarely use a web browser as a client, so I hadnt noticed this before.

After many years of using Firefox I was quite surprised to find this messages:


```
Codec not supported: ogg
```

and

```
Codec not supported: opus
```

**Comment by BlackLight** [comment]
Any progress on this issue? From what I understand the possible solutions would include:

1. Piggyback on `opus.js` - which I guess would only solve the problem for opus but not ogg?
2. Make a custom WASM wrapper
3. Wait for the webcodecs API to be released

Is my understanding correct that the `opus.js` way may be the favourite one?

Btw if the problem is only with raw opus and not with ogg opus then why do I get `Codec not supported: ogg` as well?

**Comment by BlackLight** [comment]
> It's some "opus support" or "ogg support" missing from Snapcast . Am I correct ?

It's not on Snapcast, technically speaking. If you configure `codec=opus` on Snapserver and then you play from the Snapclient service or from the Android app, everything goes well.

That's because those applications use the native Opus decoders.

The problem is only with the Web side, as raw Opus (apparently) isn't natively supported, thus support can be added either through an external library, or by sideloading the codec through a WASM wrapper.

**Comment by kingosticks** [comment]
https://github.com/eshaz/wasm-audio-decoders might be useful in solving this and https://github.com/badaix/snapweb/issues/14

**Comment by badaix** [comment]
Interesting, thanks for the link @kingosticks
